### PR TITLE
fix static method reference non-static

### LINKED ISSUE #37830: Find all references counts references for static member and regular member with same name
From https://github.com/microsoft/vscode/issues/94429

**TypeScript Version:**  3.9-beta

**Search Terms:**
- find all references
- f12

**Code**
For the code:

```ts
class Y{
	foo(){}
}
class X extends Y{
	static foo(){}
}
```

1. Run `find all references` on either instance of `foo`

**Bug:**
Two references are returned for both cases of `foo`. There should only be a single reference (itself) in both of these cases:

### PR DESCRIPTION
**Pull request by ShuiRuTian**
Fixes #37830

### CODE REVIEW

**Comment by ShuiRuTian** [comment]
Not sure that the test behavior is correct, please check them.

**Comment by DanielRosenwasser** [comment]
Just as a heads up, your commits don't seem to be associated with your GitHub account. While this isn't technically a problem, you might care if you want more appropriate attribution. You can either [make sure your GitHub account is associated with the email address](https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/adding-an-email-address-to-your-github-account) that you're using for your commits, or [rebase and amend your commits](https://stackoverflow.com/a/31669789) to fix the author name and email.

If you don't care about any of this stuff, we can start reviewing.

File: src/services/findAllReferences.ts [hunk 1]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
             readonly symbol: Symbol;
             readonly kind: NodeEntryKind | undefined;
         }
+
+        function isStatic(symbol: Symbol): boolean{
+            if(!symbol?.valueDeclaration) {return false;}
+            if(symbol.valueDeclaration?.modifierFlagsCache) {return !!((symbol.valueDeclaration.modifierFlagsCache & ModifierFlags.Static));}
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2036]: Can you update your changes to reflect the style of the rest of the file (space after keywords etc.)? The editor's default formatting works well here.

**Comment by ShuiRuTian** [comment]
> Just as a heads up, your commits don't seem to be associated with your GitHub account. While this isn't technically a problem, you might care if you want more appropriate attribution. You can either [make sure your GitHub account is associated with the email address](https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/adding-an-email-address-to-your-github-account) that you're using for your commits, or [rebase and amend your commits](https://stackoverflow.com/a/31669789) to fix the author name and email.
> 
> If you don't care about any of this stuff, we can start reviewing.

Oh, a big thank you! I did not notice it!

File: src/services/findAllReferences.ts (hunk 1 above)
Comment (ShuiRuTian) [reply src/services/findAllReferences.ts]: fixed.

**Comment by ShuiRuTian** [comment]
Friendly ping @DanielRosenwasser ~

**Comment by ShuiRuTian** [comment]
Friendly ping @DanielRosenwasser ~

File: src/services/findAllReferences.ts [hunk 2]
```diff
@@ -1888,14 +1888,30 @@ namespace ts.FindAllReferences {
...
         function populateSearchSymbolSet(symbol: Symbol, location: Node, checker: TypeChecker, isForRename: boolean, providePrefixAndSuffixText: boolean, implementations: boolean): Symbol[] {
             const result: Symbol[] = [];
+            const isSymbolStatic = isStatic(symbol);
             forEachRelatedSymbol<void>(symbol, location, checker, isForRename, !(isForRename && providePrefixAndSuffixText),
-                (sym, root, base) => { result.push(base || root || sym); },
-                /*allowBaseTypes*/ () => !implementations);
+                (sym, root, base) => {
+                    // static method/property and instance method/property might have the same name. Only include static or only include instance.
+                    if (base) {
+                        if ((isSymbolStatic && !isStatic(base)) || (!isSymbolStatic && isStatic(base))) {
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:1896]:
```suggestion
                        if (isSymbolStatic !== !isStatic(base)) {
```

File: src/services/findAllReferences.ts [hunk 3]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
             readonly symbol: Symbol;
             readonly kind: NodeEntryKind | undefined;
         }
+
+        function isStatic(symbol: Symbol): boolean {
+            if (!symbol?.valueDeclaration) { return false; }
+            if (symbol.valueDeclaration?.modifierFlagsCache) { return !!((symbol.valueDeclaration.modifierFlagsCache & ModifierFlags.Static)); }
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2036]: Why does this need a `?.`?

File: src/services/findAllReferences.ts [hunk 4]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
             readonly symbol: Symbol;
             readonly kind: NodeEntryKind | undefined;
         }
+
+        function isStatic(symbol: Symbol): boolean {
+            if (!symbol?.valueDeclaration) { return false; }
+            if (symbol.valueDeclaration?.modifierFlagsCache) { return !!((symbol.valueDeclaration.modifierFlagsCache & ModifierFlags.Static)); }
+            else { return !!(symbol.valueDeclaration.modifiers?.some((modifier) => modifier.kind & ModifierFlags.Static)); };
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2037]:
```suggestion
            return !!forEach(symbol.valueDeclaration.modifiers, modifier => modifier.kind & ModifierFlags.Static));
```

File: src/services/findAllReferences.ts [hunk 5]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
...
+            const cbSymbol = (sym: Symbol, rootSymbol: Symbol, baseSymbol: Symbol | undefined, kind: NodeEntryKind): RelatedSymbol | undefined => {
+                // check whether the symbol used to search itself is just the searched one.
+                if (baseSymbol) {
+                    // static method/property and instance method/property might have the same name. Only check static or only check instance.
+                    if ((isReferenceSymbolStatic && !isStatic(baseSymbol)) || (!isReferenceSymbolStatic && isStatic(baseSymbol))) {
+                        baseSymbol = undefined;
+                    }
+                }
+
+                return search.includes(baseSymbol || rootSymbol || sym)
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2052]: `includes` is too new, use one of the helper utilities in `core.ts`

File: src/services/findAllReferences.ts [hunk 6]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
...
+        function isStatic(symbol: Symbol): boolean {
+            if (!symbol?.valueDeclaration) { return false; }
+            if (symbol.valueDeclaration?.modifierFlagsCache) { return !!((symbol.valueDeclaration.modifierFlagsCache & ModifierFlags.Static)); }
+            else { return !!(symbol.valueDeclaration.modifiers?.some((modifier) => modifier.kind & ModifierFlags.Static)); };
+        }
+
         function getRelatedSymbol(search: Search, referenceSymbol: Symbol, referenceLocation: Node, state: State): RelatedSymbol | undefined {
             const { checker } = state;
+            const isReferenceSymbolStatic = isStatic(referenceSymbol);
+            const cbSymbol = (sym: Symbol, rootSymbol: Symbol, baseSymbol: Symbol | undefined, kind: NodeEntryKind): RelatedSymbol | undefined => {
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2043]: Switch to function declarations, not `const`s

File: src/services/findAllReferences.ts [hunk 7]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
...
+                return search.includes(baseSymbol || rootSymbol || sym)
+                    // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
+                    ? {
+                        symbol: rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym,
+                        kind
+                    }
+                    : undefined;
+            };
+            const allowBaseTypes = (rootSymbol: Symbol) =>
+                !(search.parents?.every(parent => !explicitlyInheritsFrom(rootSymbol.parent!, parent, state.inheritsFromCache, checker)));
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2061]:
```suggestion
                forEach(search.parents, parent => !explicitlyInheritsFrom(rootSymbol.parent!, parent, state.inheritsFromCache, checker));
```

File: src/services/findAllReferences.ts [hunk 8]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
...
             return forEachRelatedSymbol(referenceSymbol, referenceLocation, checker, /*isForRenamePopulateSearchSymbolSet*/ false,
                 /*onlyIncludeBindingElementAtReferenceLocation*/ state.options.use !== FindReferencesUse.Rename || !!state.options.providePrefixAndSuffixTextForRename,
-                (sym, rootSymbol, baseSymbol, kind): RelatedSymbol | undefined => search.includes(baseSymbol || rootSymbol || sym)
-                    // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
-                    ? { symbol: rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym, kind }
-                    : undefined,
-                /*allowBaseTypes*/ rootSymbol =>
-                    !(search.parents && !search.parents.some(parent => explicitlyInheritsFrom(rootSymbol.parent!, parent, state.inheritsFromCache, checker))));
+                cbSymbol,
+                /*allowBaseTypes*/allowBaseTypes);
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2066]:
```suggestion
                /*allowBaseTypes*/ allowBaseTypes);
```

File: src/services/findAllReferences.ts [hunk 9]
```diff
@@ -2014,16 +2030,40 @@ namespace ts.FindAllReferences {
...
+                if (baseSymbol) {
+                    // static method/property and instance method/property might have the same name. Only check static or only check instance.
+                    if ((isReferenceSymbolStatic && !isStatic(baseSymbol)) || (!isReferenceSymbolStatic && isStatic(baseSymbol))) {
+                        baseSymbol = undefined;
+                    }
+                }
+
+                return search.includes(baseSymbol || rootSymbol || sym)
+                    // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
+                    ? {
```
Comment (DanielRosenwasser) [inline src/services/findAllReferences.ts:2054]: This is too hard to read. Just switch to

```ts
if (!contains(baseSymbol || rootSymbol || sym)) {
    return undefined;
}

return {
    rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym,
    kind
};

File: src/services/findAllReferences.ts (hunk 7 above)
Comment (ShuiRuTian) [reply src/services/findAllReferences.ts:2061]: should allowBaseTypes be converted to function?

File: src/services/findAllReferences.ts (hunk 5 above)
Comment (ShuiRuTian) [reply src/services/findAllReferences.ts]: Oh, the type of search is Search rather than Array.

**Comment by ShuiRuTian** [comment]
Isn't it finished? Why it says CI build is in progress?

File: src/compiler/core.ts [hunk 10]
```diff
@@ -149,16 +149,17 @@ namespace ts {
...
      * If no such value is found, the callback is applied to each element of array and `true` is returned.
+     * If `array` is undefined, return false.
      */
     export function every<T>(array: readonly T[] | undefined, callback: (element: T, index: number) => boolean): boolean {
-        if (array) {
-            for (let i = 0; i < array.length; i++) {
-                if (!callback(array[i], i)) {
-                    return false;
-                }
+        if (!array) {
```
Comment (sheetalkamat) [inline src/compiler/core.ts:155]: interesting this should be !length(array) instead

File: src/services/findAllReferences.ts [hunk 11]
```diff
@@ -2014,16 +2030,41 @@ namespace ts.FindAllReferences {
...
+        function isStatic(symbol: Symbol): boolean {
+            if (!symbol.valueDeclaration) { return false; }
+            if (symbol.valueDeclaration?.modifierFlagsCache) { return !!((symbol.valueDeclaration.modifierFlagsCache & ModifierFlags.Static)); }
+            return !!forEach(symbol.valueDeclaration.modifiers, modifier => modifier.kind & ModifierFlags.Static);
+        }
+
         function getRelatedSymbol(search: Search, referenceSymbol: Symbol, referenceLocation: Node, state: State): RelatedSymbol | undefined {
             const { checker } = state;
+            const isReferenceSymbolStatic = isStatic(referenceSymbol);
+            function cbSymbol(sym: Symbol, rootSymbol: Symbol, baseSymbol: Symbol | undefined, kind: NodeEntryKind): RelatedSymbol | undefined {
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2043]: Can you please keep this inlined or move it at the end instead.

File: src/services/findAllReferences.ts [hunk 12]
```diff
@@ -2014,16 +2030,41 @@ namespace ts.FindAllReferences {
             readonly symbol: Symbol;
             readonly kind: NodeEntryKind | undefined;
         }
+
+        function isStatic(symbol: Symbol): boolean {
+            if (!symbol.valueDeclaration) { return false; }
+            if (symbol.valueDeclaration?.modifierFlagsCache) { return !!((symbol.valueDeclaration.modifierFlagsCache & ModifierFlags.Static)); }
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2036]: I think you want to use `getEffectiveModifierFlags`?

File: src/services/findAllReferences.ts [hunk 13]
```diff
@@ -2014,16 +2030,41 @@ namespace ts.FindAllReferences {
...
+        }
+
         function getRelatedSymbol(search: Search, referenceSymbol: Symbol, referenceLocation: Node, state: State): RelatedSymbol | undefined {
             const { checker } = state;
+            const isReferenceSymbolStatic = isStatic(referenceSymbol);
+            function cbSymbol(sym: Symbol, rootSymbol: Symbol, baseSymbol: Symbol | undefined, kind: NodeEntryKind): RelatedSymbol | undefined {
+                // check whether the symbol used to search itself is just the searched one.
+                if (baseSymbol) {
+                    // static method/property and instance method/property might have the same name. Only check static or only check instance.
+                    if ((isReferenceSymbolStatic && !isStatic(baseSymbol)) || (!isReferenceSymbolStatic && isStatic(baseSymbol))) {
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2047]: why not just check isReferecedSymbolStatic !== isStatic(baseSymbol)

File: src/services/findAllReferences.ts [hunk 14]
```diff
@@ -2014,16 +2030,41 @@ namespace ts.FindAllReferences {
...
+                    return undefined;
+                }
+                // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
+                return {
+                    symbol: rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym,
+                    kind
+                };
+            };
+            const allowBaseTypes = (rootSymbol: Symbol) =>
+                !every(search.parents, parent => !explicitlyInheritsFrom(rootSymbol.parent!, parent, state.inheritsFromCache, checker));
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2062]: use `search.parents?.every` instead and revert changes in core

File: src/compiler/core.ts (hunk 10 above)
Comment (ShuiRuTian) [reply src/compiler/core.ts:155]: A little confused, another review says "revert changes in core", so this is also reverted?

File: src/services/findAllReferences.ts (hunk 13 above)
Comment (ShuiRuTian) [reply src/services/findAllReferences.ts]: Whooops, I forget to check other codes, sorry for that!

File: src/services/findAllReferences.ts [hunk 15]
```diff
@@ -1888,14 +1888,30 @@ namespace ts.FindAllReferences {
...
         function populateSearchSymbolSet(symbol: Symbol, location: Node, checker: TypeChecker, isForRename: boolean, providePrefixAndSuffixText: boolean, implementations: boolean): Symbol[] {
             const result: Symbol[] = [];
+            const isSymbolStatic = isStatic(symbol);
             forEachRelatedSymbol<void>(symbol, location, checker, isForRename, !(isForRename && providePrefixAndSuffixText),
-                (sym, root, base) => { result.push(base || root || sym); },
-                /*allowBaseTypes*/ () => !implementations);
+                (sym, root, base) => {
+                    // static method/property and instance method/property might have the same name. Only include static or only include instance.
+                    if (base) {
+                        if (isSymbolStatic !== isStatic(base)) {
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:1896]:
```suggestion
                        if (isStatic(symbol) !== isStatic(base)) {
```

File: src/services/findAllReferences.ts [hunk 16]
```diff
@@ -1888,14 +1888,30 @@ namespace ts.FindAllReferences {
         // This is not needed when searching for re-exports.
         function populateSearchSymbolSet(symbol: Symbol, location: Node, checker: TypeChecker, isForRename: boolean, providePrefixAndSuffixText: boolean, implementations: boolean): Symbol[] {
             const result: Symbol[] = [];
+            const isSymbolStatic = isStatic(symbol);
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:1891]:
```suggestion
```

File: src/services/findAllReferences.ts [hunk 17]
```diff
@@ -2014,16 +2030,41 @@ namespace ts.FindAllReferences {
...
+            const modifierFlags = getEffectiveModifierFlags(symbol.valueDeclaration);
+            return !!(modifierFlags & ModifierFlags.Static);
+        }
+
         function getRelatedSymbol(search: Search, referenceSymbol: Symbol, referenceLocation: Node, state: State): RelatedSymbol | undefined {
             const { checker } = state;
+            const isReferenceSymbolStatic = isStatic(referenceSymbol);
+            const allowBaseTypes = (rootSymbol: Symbol) =>
+                !(search.parents?.every(parent => !explicitlyInheritsFrom(rootSymbol.parent!, parent, state.inheritsFromCache, checker)));
+
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2045]:
```suggestion
```

File: src/services/findAllReferences.ts [hunk 18]
```diff
@@ -2014,16 +2030,41 @@ namespace ts.FindAllReferences {
...
+                }
+                if (!search.includes(baseSymbol || rootSymbol || sym)) {
+                    return undefined;
+                }
+                // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
+                return {
+                    symbol: rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym,
+                    kind
+                };
+            };
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2067]:
```suggestion
                (sym, rootSymbol, baseSymbol, kind): RelatedSymbol | undefined => {
                    // check whether the symbol used to search itself is just the searched one.
                    if (baseSymbol) {
                        // static method/property and instance method/property might have the same name. Only check static or only check instance.
                        if (isStatic(referenceSymbol) !== isStatic(baseSymbol)) {
                            baseSymbol = undefined;
                        }
                    }
                    return search.includes(baseSymbol || rootSymbol || sym)
                        // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
                        ? { symbol: rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym, kind }
                        : undefined
                },
                /*allowBaseTypes*/ rootSymbol =>
                    !(search.parents && !search.parents.some(parent => explicitlyInheritsFrom(rootSymbol.parent!, parent, state.inheritsFromCache, checker)))
            );
```

File: src/services/findAllReferences.ts [hunk 19]
```diff
@@ -2014,16 +2029,33 @@ namespace ts.FindAllReferences {
...
+                    if (baseSymbol) {
+                        // static method/property and instance method/property might have the same name. Only check static or only check instance.
+                        if (isStatic(referenceSymbol) !== isStatic(baseSymbol)) {
+                            baseSymbol = undefined;
+                        }
+                    }
+                    return search.includes(baseSymbol || rootSymbol || sym)
+                        // For a base type, use the symbol for the derived type. For a synthetic (e.g. union) property, use the union symbol.
+                        ? { symbol: rootSymbol && !(getCheckFlags(sym) & CheckFlags.Synthetic) ? rootSymbol : sym, kind }
+                        : undefined
```
Comment (sheetalkamat) [inline src/services/findAllReferences.ts:2054]:
```suggestion
                        : undefined;
```

**Comment by sheetalkamat** [comment]
@ShuiRuTian thank you for working on this

**Pull request merged**